File: data/benchmark0/cpp/reverse.cpp
/*
reverse an array in-place
- the c++ implementation uses two raw pointers advancing toward each other from both ends
- a naive rust implementation would keep pointer arithmetic with unsafe
- a good rust implementation uses two indices on a &mut [i32] slice
*/

void reverse(int* a, int n) {
    int* lo = a;
    int* hi = a + n - 1;
    while (lo < hi) {
        int tmp = *lo;
        *lo = *hi;
        *hi = tmp;
        lo++;
        hi--;
    }
}

File: data/benchmark0/rust/reverse.rs
pub fn reverse(a: &mut [i32]) {
    if a.len() == 0 { return; }
    let mut lo = 0usize;
    let mut hi = a.len() - 1;
    let n = a.len();
    while lo < hi {
        let hi = n - 1 - lo;
        let tmp = a[lo];
        a[lo] = a[hi];
        a[hi] = tmp;
        lo += 1;
    }
}
